File: apps/web/src/app/workspace.ts
import {
  addDays,
  ancestorFolders,
  basename,
  type DailyNoteResponse,
  dirname,
  ensureMarkdownExtension,
  isHiddenPath,
  type LocalDate,
  normalizePath,
  resolveWikiLink,
  type ServerEventOf,
  stem,
  today,
  toISODate,
  type VaultTreeResponse,
} from "@ddl/core";
import type { DaemonClient } from "../api/client";
import { ConflictError, errorMessage } from "../api/errors";
import { adjacentDailyTarget, dailyPathFor } from "../features/daily/daily-nav";
import { AnnotationSync } from "../features/editor/annotation-sync";
import { EditorController } from "../features/editor/editor-controller";
import { PresenceReporter } from "../features/editor/presence";
import { WordCounter } from "../features/editor/word-counter";
import { editorConfigFrom } from "../features/settings/theme";
import { afterNextPaint } from "../lib/idle";
import {
  type PerfMetric,
  perfAnnotate,
  perfCancel,
  perfEndAfterPaint,
  perfStart,
  recordMeasure,
} from "../perf/perf";
import { NotesController } from "../state/notes-controller";
import { setSaveState, setWordCount } from "../state/notes-store";
import { getSettings } from "../state/settings-store";
import { placeInTabs, removeFromTabs, renameInTabs, useTabsStore } from "../state/tabs-store";
import { toast } from "../state/toast-store";
import { ui } from "../state/ui-store";
import { vaultActions } from "../state/vault-store";
import type { AgentActions } from "./agent-actions";

export interface OpenOptions {
  newTab?: boolean;
  /** 0-based line to scroll to. */
  line?: number;
  focus?: boolean;
  metric?: PerfMetric;
}

const MAX_CACHED_NOTES = 24;
const INVALID_NAME_RE = /[\\/:*?"<>|#^[\]]/;

export function validateNoteName(name: string): string | null {
  const trimmed = name.trim();
  if (!trimmed) return "Name can't be empty";
  if (trimmed.startsWith(".")) return "Name can't start with a dot";
  if (INVALID_NAME_RE.test(trimmed)) return "Name can't contain any of \\ / : * ? \" < > | # ^ [ ]";
  return null;
}

function openExternal(url: string): void {
  window.open(url, "_blank", "noopener,noreferrer");
}

/**
 * The note-side application controller: tabs, navigation (incl. daily notes), CRUD, external
 * changes and flush-on-leave. It drives the editor imperatively; React only renders state.
 */
export class Workspace {
  readonly client: DaemonClient;
  readonly notes: NotesController;
  readonly editor: EditorController;
  readonly annotations: AnnotationSync;
  private readonly agent: AgentActions;
  private readonly presence: PresenceReporter;
  private readonly words: WordCounter;
  private readonly recent: string[] = [];
  private readonly errorToasted = new Set<string>();
  private navToken = 0;
  /** Where the latest navigation is going while its note loads. */
  private navTarget: string | null = null;
  private treeRefresh: ReturnType<typeof setTimeout> | undefined;
  private mountWaiters: Array<() => void> = [];

  constructor(client: DaemonClient, agent: AgentActions) {
    this.client = client;
    this.agent = agent;
    this.notes = new NotesController({
      client,
      hooks: {
        readLive: (path) => this.editor.readLive(path),
        applyRemote: (path, content) => this.editor.applyRemote(path, content),
        onSaveState: (path, state) => {
          setSaveState(path, state);
          if (state === "saved") this.errorToasted.delete(path);
        },
        onConflictCopy: (path, copyPath) => this.onConflictCopy(path, copyPath),
        onRemoteDelete: (path, restored) => this.onRemoteDelete(path, restored),
        onSaveError: (path, error) => this.onSaveError(path, error),
        pathExists: (path) => vaultActions.has(path),
      },
    });
    this.editor = new EditorController(
      {
        contentOf: (path) => this.notes.serverContent(path),
        onLocalEdit: (path) => {
          this.notes.markDirty(path);
          this.annotations.onDocumentEdited();
          this.presence.onEdit(path);
          this.words.schedule();
        },
        onDocumentReplaced: () => {
          this.annotations.recompute();
          this.words.schedule(0);
        },
        onCursorLine: (path, line) => this.presence.onCursorLine(path, line),
        onAnnotationClick: (annotation) =>
          this.agent.openTaskThread(annotation.id, annotation.threadId),
        onWikiLinkClick: (target, newPane) => void this.openWikiLink(target, newPane),
        onExternalLinkClick: openExternal,
        onSaveRequested: (path) => void this.notes.flush(path),
        beforeDeactivate: (path) => void this.notes.flush(path),
        canEvict: (path) =>
          !useTabsStore.getState().tabs.includes(path) && !this.notes.isBusy(path),
      },
      editorConfigFrom(getSettings()),
    );
    this.annotations = new AnnotationSync(this.editor);
    this.presence = new PresenceReporter((notePath, line) =>
      client.send({ type: "editor.activity", notePath, line }),
    );
    this.words = new WordCounter(() => this.editor.getDocument(), setWordCount);
  }

  get activePath(): string | null {
    return useTabsStore.getState().active;
  }

  mountEditor(parent: HTMLElement): void {
    this.editor.mount(parent);
    const waiters = this.mountWaiters;
    this.mountWaiters = [];
    for (const resolve of waiters) resolve();
  }

  unmountEditor(): void {
    this.editor.unmount();
  }

  whenEditorMounted(): Promise<void> {
    if (this.editor.mounted) return Promise.resolve();
    return new Promise((resolve) => this.mountWaiters.push(resolve));
  }

  // ── Opening notes ──────────────────────────────────────────────────────

  /** Starts a navigation, superseding any that is still loading. `target`: where it is going. */
  private beginNavigation(target: string | null): number {
    this.navTarget = target;
    return ++this.navToken;
  }

  private endNavigation(token: number): void {
    if (token === this.navToken) this.navTarget = null;
  }

  async openNote(path: string, options: OpenOptions = {}): Promise<boolean> {
    const token = this.beginNavigation(path);
    if (!this.notes.has(path)) {
      try {
        await this.notes.load(path);
      } catch (error) {
        if (options.metric) perfCancel(options.metric);
        if (token === this.navToken) {
          toast({
            kind: "error",
            title: `Couldn't open “${stem(path)}”`,
            body: errorMessage(error),
          });
        }
        this.endNavigation(token);
        return false;
      }
      if (token !== this.navToken) {
        if (options.metric) perfCancel(options.metric);
        return false;
      }
    }
    this.endNavigation(token);
    this.activate(path, options);
    return true;
  }

  /** Synchronous switch to an already-loaded note. */
  activate(path: string, options: OpenOptions = {}): void {
    const state = useTabsStore.getState();
    const next = placeInTabs(state, path, options.newTab ?? false);
    this.editor.show(path);
    if (next.tabs !== state.tabs || next.active !== state.active) useTabsStore.setState(next);
    this.afterActivate(path, options);
  }

  activateTab(path: string, start?: number): void {
    // The click wins over a note still loading from an earlier navigation.
    this.beginNavigation(null);
    if (useTabsStore.getState().active === path) return;
    perfStart("tab:switch", start);
    this.activate(path, { metric: "tab:switch" });
  }

  closeTab(path: string): void {
    const state = useTabsStore.getState();
    const next = removeFromTabs(state, path);
    if (next === state) return;
    void this.notes.flush(path);
    const wasActive = state.active === path;
    if (wasActive) this.editor.show(next.active);
    useTabsStore.setState(next);
    if (!wasActive) return;
    if (next.active) this.afterActivate(next.active, {});
    else this.onNoActiveNote();
  }

  closeActiveTab(): void {
    const active = this.activePath;
    if (active) this.closeTab(active);
  }

  private afterActivate(path: string, options: OpenOptions): void {
    if (options.line !== undefined) this.editor.scrollToLine(options.line);
    if (options.focus !== false) this.editor.focus();
    this.annotations.setActive(path);
    this.words.schedule(0);
    this.agent.refreshRecords(path);
    const ancestors = ancestorFolders(path);
    if (ancestors.length > 0) ui.expandAll(ancestors);
    this.touch(path);
    if (options.metric) perfEndAfterPaint(options.metric);
  }

  private onNoActiveNote(): void {
    this.annotations.setActive(null);
    this.presence.reset();
    setWordCount(null);
  }

  /** Keeps a bounded set of recently used notes loaded (instant back-navigation). */
  private touch(path: string): void {
    const index = this.recent.indexOf(path);
    if (index !== -1) this.recent.splice(index, 1);
    this.recent.unshift(path);
    const tabs = useTabsStore.getState().tabs;
    for (let i = this.recent.length - 1; i >= MAX_CACHED_NOTES; i--) {
      const candidate = this.recent[i]!;
      if (
        tabs.includes(candidate) ||
        this.notes.isBusy(candidate) ||
        candidate === this.activePath
      ) {
        continue;
      }
      this.recent.splice(i, 1);
      this.notes.forget(candidate);
      this.editor.forget(candidate);
    }
  }

  /** Drops a note from the caches (test hook for measuring uncached opens). */
  evict(path: string): void {
    if (useTabsStore.getState().tabs.includes(path) || this.notes.isBusy(path)) return;
    const index = this.recent.indexOf(path);
    if (index !== -1) this.recent.splice(index, 1);
    this.notes.forget(path);
    this.editor.forget(path);
  }

  async openWikiLink(target: string, newTab: boolean): Promise<void> {
    const resolved = resolveWikiLink(target, vaultActions.files());
    if (resolved) {
      await this.openNote(resolved, { newTab });
      return;
    }
    let path: string;
    try {
      path = ensureMarkdownExtension(normalizePath(target));
    } catch {
      toast({ kind: "error", title: "Invalid link target", body: target });
      return;
    }
    await this.createNoteAt(path, { newTab, focusTitle: false });
  }

  // ── Daily notes ────────────────────────────────────────────────────────

  openToday(start?: number): Promise<void> {
    return this.openDaily(today(), "daily:open", start);
  }

  openTomorrow(): Promise<void> {
    return this.openDaily(addDays(today(), 1), "daily:open");
  }

  /** Opens (creating from the template if needed — server side) the daily note for `date`. */
  async openDaily(
    date: LocalDate,
    metric: PerfMetric = "daily:open",
    start?: number,
  ): Promise<void> {
    perfStart(metric, start);
    const path = dailyPathFor(date, getSettings().dailyNotes);
    if (path && this.notes.has(path)) {
      this.beginNavigation(null);
      perfAnnotate(metric, { cached: true });
      this.activate(path, { metric });
      return;
    }
    perfAnnotate(metric, { cached: false });
    const token = this.beginNavigation(path);
    try {
      const note = await this.client.getDailyNote(toISODate(date), true);
      if (token !== this.navToken) {
        perfCancel(metric);
        return;
      }
      this.endNavigation(token);
      this.adoptDaily(note);
      this.activate(note.path, { metric });
    } catch (error) {
      this.endNavigation(token);
      perfCancel(metric);
      toast({ kind: "error", title: "Couldn't open the daily note", body: errorMessage(error) });
    }
  }

  adoptDaily(note: DailyNoteResponse): void {
    this.notes.adopt(note);
    if (!vaultActions.has(note.path)) vaultActions.addFile(note.path, note.version);
  }

  /** Startup: show today's note and record `app:interactive` once it is on screen. */
  async showStartupNote(note: DailyNoteResponse): Promise<void> {
    this.adoptDaily(note);
    await this.whenEditorMounted();
    if (this.activePath === null) this.activate(note.path, {});
    afterNextPaint(() => recordMeasure("app:interactive", 0, performance.now()));
  }

  openAdjacentDaily(direction: -1 | 1, start?: number): Promise<boolean> {
    const settings = getSettings().dailyNotes;
    // Repeated keys continue from the note still loading, instead of re-picking the same one.
    const from = this.navTarget ?? this.activePath;
    const target = adjacentDailyTarget(vaultActions.files(), from, direction, settings);
    if (!target) {
      toast({
        kind: "info",
        title: direction === -1 ? "No previous daily note" : "No next daily note",
        timeoutMs: 2500,
      });
      return Promise.resolve(false);
    }
    const metric: PerfMetric = direction === -1 ? "daily:prev" : "daily:next";
    perfStart(metric, start, { cached: this.notes.has(target.path) });
    return this.openNote(target.path, { metric });
  }

  // ── Create / rename / delete ───────────────────────────────────────────

  private uniquePath(folder: string, base: string, extension = ".md"): string {
    const prefix = folder ? `${folder}/` : "";
    for (let n = 0; ; n++) {
      const candidate = `${prefix}${n === 0 ? base : `${base} ${n}`}${extension}`;
      if (!vaultActions.has(candidate)) return candidate;
    }
  }

  async createNote(options: { folder?: string; name?: string; newTab?: boolean } = {}) {
    const name = options.name?.trim();
    if (name) {
      let path: string;
      try {
        path = ensureMarkdownExtension(
          normalizePath(options.folder ? `${options.folder}/${name}` : name),
        );
      } catch {
        toast({ kind: "error", title: "Invalid note name", body: name });
        return null;
      }
      const invalid = validateNoteName(stem(path));
      if (invalid) {
        toast({ kind: "error", title: "Invalid note name", body: invalid });
        return null;
      }
      return this.createNoteAt(path, { newTab: options.newTab ?? true, focusTitle: false });
    }
    const path = this.uniquePath(options.folder ?? "", "Untitled");
    return this.createNoteAt(path, { newTab: options.newTab ?? true, focusTitle: true });
  }

  private async createNoteAt(
    path: string,
    options: { newTab: boolean; focusTitle: boolean },
  ): Promise<string | null> {
    if (vaultActions.has(path)) {
      await this.openNote(path, { newTab: options.newTab });
      return path;
    }
    try {
      const response = await this.client.writeNote(path, { content: "", baseVersion: null });
      vaultActions.addFile(response.path, response.version);
      this.notes.adopt({
        path: response.path,
        content: "",
        version: response.version,
        mtime: response.mtime,
      });
      this.activate(response.path, { newTab: options.newTab, focus: !options.focusTitle });
      if (options.focusTitle) ui.set({ titleFocus: response.path });
      return response.path;
    } catch (error) {
      if (error instanceof ConflictError && error.current) {
        this.notes.adopt(error.current);
        vaultActions.addFile(path);
        this.activate(path, { newTab: options.newTab });
        return path;
      }
      toast({ kind: "error", title: "Couldn't create the note", body: errorMessage(error) });
      return null;
    }
  }

  async createFolder(parent = ""): Promise<string | null> {
    const path = this.uniquePath(parent, "Untitled folder", "");
    try {
      await this.client.createFolder(path);
    } catch (error) {
      toast({ kind: "error", title: "Couldn't create the folder", body: errorMessage(error) });
      return null;
    }
    vaultActions.addFolder(path);
    if (parent) ui.setExpanded(parent, true);
    ui.set({ renaming: path });
    return path;
  }

  /** Renames the note's file stem (title rename). */
  renameNoteTitle(path: string, title: string): Promise<boolean> {
    const invalid = validateNoteName(title);
    if (invalid) {
      toast({ kind: "error", title: "Can't rename", body: invalid });
      return Promise.resolve(false);
    }
    const folder = dirname(path);
    const target = `${folder ? `${folder}/` : ""}${title.trim()}.md`;
    return this.renamePath(path, target);
  }

  /** Renames a file or folder in place (explorer inline rename). */
  renameEntry(path: string, name: string): Promise<boolean> {
    const invalid = validateNoteName(name);
    if (invalid) {
      toast({ kind: "error", title: "Can't rename", body: invalid });
      return Promise.resolve(false);
    }
    const folder = dirname(path);
    const isFolder = vaultActions.isFolder(path);
    const leaf = isFolder ? name.trim() : ensureMarkdownExtension(name.trim());
    return this.renamePath(path, `${folder ? `${folder}/` : ""}${leaf}`);
  }

  async renamePath(from: string, to: string): Promise<boolean> {
    if (from === to) return true;
    if (vaultActions.has(to)) {
      toast({ kind: "error", title: "Can't rename", body: `“${basename(to)}” already exists` });
      return false;
    }
    const affected = this.notes.paths().filter((p) => p === from || p.startsWith(`${from}/`));
    await Promise.all(affected.map((p) => this.notes.flush(p)));
    try {
      await this.client.renamePath(from, to);
    } catch (error) {
      toast({ kind: "error", title: "Couldn't rename", body: errorMessage(error) });
      return false;
    }
    vaultActions.rename(from, to);
    this.notes.rename(from, to);
    this.editor.rename(from, to);
    useTabsStore.setState(renameInTabs(useTabsStore.getState(), from, to));
    for (let i = 0; i < this.recent.length; i++) {
      const p = this.recent[i]!;
      if (p === from) this.recent[i] = to;
      else if (p.startsWith(`${from}/`)) this.recent[i] = `${to}${p.slice(from.length)}`;
    }
    for (const p of affected) {
      this.agent.forgetRecords(p);
    }
    const active = this.activePath;
    if (active) {
      this.agent.refreshRecords(active, true);
      this.annotations.setActive(active);
    }
    return true;
  }

  requestDelete(path: string): void {
    const isFolder = vaultActions.isFolder(path);
    const name = isFolder ? basename(path) : stem(path);
    ui.confirm({
      title: `Delete “${name}”?`,
      message: isFolder
        ? "The folder and everything in it will be deleted. This can't be undone."
        : "The note will be deleted. This can't be undone.",
      confirmLabel: "Delete",
      danger: true,
      onConfirm: () => void this.deletePath(path),
    });
  }

  async deletePath(path: string): Promise<boolean> {
    const isFolder = vaultActions.isFolder(path);
    try {
      if (isFolder) await this.client.deleteFolder(path);
      else await this.client.deleteNote(path);
    } catch (error) {
      toast({ kind: "error", title: "Couldn't delete", body: errorMessage(error) });
      return false;
    }
    const affected = isFolder
      ? vaultActions.files().filter((p) => p.startsWith(`${path}/`))
      : [path];
    for (const p of affected) this.dropNote(p);
    vaultActions.remove(path);
    return true;
  }

  /** Removes a note that no longer exists from every cache and tab (without saving it). */
  private dropNote(path: string): void {
    this.notes.forget(path);
    this.editor.forget(path);
    this.agent.forgetRecords(path);
    const index = this.recent.indexOf(path);
    if (index !== -1) this.recent.splice(index, 1);
    this.closeTab(path);
  }

  // ── Remote changes ─────────────────────────────────────────────────────

  handleVaultChanged(event: ServerEventOf<"vault.changed">): void {
    if (event.clientId !== undefined && event.clientId === this.client.clientId) return;
    for (const change of event.changes) {
      if (isHiddenPath(change.path)) continue;
      if (change.kind === "deleted") {
        const nested = this.notes.paths().filter((p) => p.startsWith(`${change.path}/`));
        for (const p of [change.path, ...nested]) this.notes.handleRemoteDelete(p);
        vaultActions.remove(change.path);
        continue;
      }
      if (!vaultActions.has(change.path)) {
        if (/\.[^/]+$/.test(change.path)) vaultActions.addFile(change.path, change.version);
        else this.scheduleTreeRefresh();
      }
      if (this.notes.has(change.path))
        void this.notes.handleRemoteChange(change.path, change.version);
    }
  }

  private scheduleTreeRefresh(): void {
    clearTimeout(this.treeRefresh);
    this.treeRefresh = setTimeout(() => void this.refreshTree(), 300);
  }

  async refreshTree(): Promise<void> {
    try {
      this.applyTree(await this.client.getTree());
    } catch {
      // Keep the current tree; the next resync retries.
    }
  }

  /** Replaces the tree, keeping notes we know exist (the snapshot may predate a creation). */
  applyTree(tree: VaultTreeResponse): void {
    vaultActions.setTree(tree);
    for (const path of this.notes.paths())
      vaultActions.addFile(path, this.notes.version(path) ?? undefined);
  }

  /** After a reconnect: events may have been missed, so re-validate everything we show. */
  async resync(): Promise<void> {
    await this.refreshTree();
    for (const path of this.notes.paths()) void this.notes.handleRemoteChange(path);
  }

  private onConflictCopy(path: string, copyPath: string): void {
    vaultActions.addFile(copyPath);
    toast({
      kind: "warning",
      title: `“${stem(path)}” changed elsewhere`,
      body: `Kept your version. The other version was saved as “${stem(copyPath)}”.`,
      actionLabel: "Open",
      onClick: () => void this.openNote(copyPath, { newTab: true }),
      timeoutMs: 10_000,
    });
  }

  private onRemoteDelete(path: string, restored: boolean): void {
    if (restored) {
      vaultActions.addFile(path);
      toast({
        kind: "warning",
        title: `“${stem(path)}” was deleted elsewhere`,
        body: "Your unsaved edits were written back, so the note was restored.",
      });
      return;
    }
    const wasOpen = useTabsStore.getState().tabs.includes(path);
    this.dropNote(path);
    if (wasOpen) toast({ kind: "info", title: `“${stem(path)}” was deleted`, timeoutMs: 3000 });
  }

  private onSaveError(path: string, error: unknown): void {
    if (this.errorToasted.has(path)) return;
    this.errorToasted.add(path);
    toast({
      kind: "error",
      title: `Couldn't save “${stem(path)}”`,
      body: `${errorMessage(error)} — retrying automatically.`,
    });
  }

  // ── Lifecycle ──────────────────────────────────────────────────────────

  /** Flushes pending edits when the window loses focus, is hidden or unloads. */
  installLifecycle(): () => void {
    const flush = () => void this.notes.flushAll();
    const flushKeepalive = () => void this.notes.flushAll({ keepalive: true });
    const onVisibility = () => {
      if (document.visibilityState === "hidden") flushKeepalive();
    };
    window.addEventListener("blur", flush);
    window.addEventListener("pagehide", flushKeepalive);
    window.addEventListener("beforeunload", flushKeepalive);
    document.addEventListener("visibilitychange", onVisibility);
    return () => {
      window.removeEventListener("blur", flush);
      window.removeEventListener("pagehide", flushKeepalive);
      window.removeEventListener("beforeunload", flushKeepalive);
      document.removeEventListener("visibilitychange", onVisibility);
    };
  }
}

File: apps/macos/Sources/DailyDoListApp/Workspace/Workspace.swift
import DailyDoListAgent
import DailyDoListClient
import DailyDoListDomain
import DailyDoListEditor
import DailyDoListModels
import DailyDoListVim
import Foundation
import Observation

/// How a note should be opened.
struct OpenOptions: Sendable {
  /// Open in a new tab instead of replacing the active tab's note.
  var newTab = false
  /// 0-based line to scroll to.
  var line: Int?
  var focusEditor = true
  /// False for Back/Forward (they walk the history instead of extending it).
  var recordHistory = true
}

/// The note-side application controller (the native counterpart of the web app's `Workspace`):
/// tabs and history, daily/weekly notes, create/rename/delete, remote changes and resync. Stores
/// hold state; this type sequences them and drives the editor imperatively.
@MainActor
@Observable
final class Workspace {
  static let maxCachedNotes = 24

  @ObservationIgnored let client: DaemonClient
  let vault: VaultStore
  let notes: NotesStore
  let tabs: TabsStore
  let editor: EditorCoordinator
  let settings: SettingsStore
  let ui: UIState
  let search: SearchModel
  /// The drawings notes embed, and drawings opened on their own.
  @ObservationIgnored let drawings: DrawingStore
  /// Bumped whenever a drawing loaded or changed (a drawing open on its own redraws).
  private(set) var drawingsRevision = 0
  /// Drawings open on their own that show their Markdown source instead of the drawing.
  var drawingSourcePaths: Set<String> = []
  /// What the orchestrator is doing (chips on the lines that woke it, header, status bar).
  let orchestrator: OrchestratorActivityStore
  @ObservationIgnored let toasts: ToastStore
  @ObservationIgnored weak var agent: AgentStore?
  @ObservationIgnored let scheduler: AppScheduler
  /// Wall clock ("today" for daily notes); injectable for tests.
  @ObservationIgnored let now: () -> Date
  /// Local vault folder when known (Reveal in Finder).
  var localVaultURL: URL?

  /// Most recently used notes, newest first (quick switcher, cache eviction).
  private(set) var recent: [String] = []
  /// Called after tabs change (persist them).
  @ObservationIgnored var onTabsChanged: (@MainActor () -> Void)?
  /// Runs an app command by id (vim's `:obcommand`); false when there's no such command.
  @ObservationIgnored var commandRunner: (@MainActor (String) -> Bool)?
  /// Opens the orchestrator's chat at a turn (its first message's id; nil: at the end).
  @ObservationIgnored var openOrchestratorTurn: (@MainActor (String?) -> Void)?

  @ObservationIgnored var navToken = 0
  /// Where the latest navigation is going while its note loads.
  @ObservationIgnored var navTarget: String?
  @ObservationIgnored var recordNotes: Set<String> = []
  @ObservationIgnored var errorToasted: Set<String> = []
  @ObservationIgnored var treeRefresh: IdleTimer!
  @ObservationIgnored var presence: PresenceReporter!
  /// Hover previews of `[[wikilinks]]` (editor tooltips, thread cards).
  @ObservationIgnored var notePreviews: NotePreviewCache!
  /// Threads loaded for the sources of a link preview (each at most once).
  @ObservationIgnored var previewThreadLoads: Set<String> = []

  init(
    client: DaemonClient, settings: SettingsStore, ui: UIState, toasts: ToastStore,
    scheduler: AppScheduler, editorController: MarkdownEditorController? = nil, vim: Vim? = nil,
    now: @escaping () -> Date = Date.init
  ) {
    self.client = client
    self.settings = settings
    self.ui = ui
    self.toasts = toasts
    self.scheduler = scheduler
    self.now = now
    vault = VaultStore(client: client)
    notes = NotesStore(client: client, scheduler: scheduler)
    tabs = TabsStore()
    search = SearchModel(client: client, scheduler: scheduler)
    drawings = DrawingStore(client: client, scheduler: scheduler)
    orchestrator = OrchestratorActivityStore(scheduler: scheduler)
    editor = EditorCoordinator(controller: editorController, scheduler: scheduler)
    treeRefresh = IdleTimer(scheduler: scheduler, delay: 0.3) { [weak self] in
      guard let self else { return }
      Task { await self.refreshTree() }
    }
    presence = PresenceReporter(scheduler: scheduler) { [client] path, line in
      Task { await client.send(.editorActivity(notePath: path, line: line)) }
    }
    notePreviews = NotePreviewCache(
      resolve: { [weak self] target in self?.resolveWikiLink(target) },
      loadedContent: { [weak self] path in
        guard let self, self.notes.has(path) else { return nil }
        return self.editor.liveText(for: path) ?? self.content(of: path)
      },
      read: { [client] path in try await client.readNote(path).content })
    notes.delegate = self
    notes.onStateChange = { [weak self] path, state in
      if state == .saved { self?.errorToasted.remove(path) }
    }
    editor.host = self
    setUpDrawings()
    orchestrator.onChipsChanged = { [weak self] notes in self?.editor.chipsDidChange(for: notes) }
    editor.controller.vim = vim
    editor.configure(settings.settings.editor)
  }

  func drawingsDidChange() {
    drawingsRevision += 1
  }

  var activePath: String? { tabs.active }

  // MARK: - Navigation bookkeeping

  /// Starts a navigation, superseding any still loading. `target`: where it is going.
  func beginNavigation(_ target: String?) -> Int {
    navTarget = target
    navToken += 1
    return navToken
  }

  func endNavigation(_ token: Int) {
    if token == navToken { navTarget = nil }
  }

  // MARK: - Opening notes

  @discardableResult
  func openNote(_ path: String, _ options: OpenOptions = OpenOptions()) async -> Bool {
    let token = beginNavigation(path)
    if !notes.has(path) {
      do {
        try await notes.load(path)
      } catch {
        if token == navToken { toasts.error("Couldn't open “\(VaultPath.stem(path))”", error) }
        endNavigation(token)
        return false
      }
      guard token == navToken else { return false }
    }
    endNavigation(token)
    activate(path, options)
    return true
  }

  /// Synchronous switch to an already-loaded note.
  func activate(_ path: String, _ options: OpenOptions = OpenOptions()) {
    switchEditor(to: path) {
      tabs.place(path, newTab: options.newTab, recordHistory: options.recordHistory)
    }
    afterActivate(path, options)
  }

  /// Tab click: wins over a note still loading from an earlier navigation.
  func activateTab(_ path: String) {
    _ = beginNavigation(nil)
    guard tabs.active != path else { return }
    activate(path)
  }

  func closeTab(_ path: String) {
    guard tabs.tabs.contains(path) else { return }
    notes.saveNow(path)
    let wasActive = tabs.active == path
    let next = tabs.close(path)
    if wasActive {
      editor.show(
        next, content: next.flatMap(content(of:)), caretAtEnd: next.map(isDailyNote) ?? false)
      if let next { afterActivate(next, OpenOptions()) } else { onNoActiveNote() }
    }
    onTabsChanged?()
  }

  func closeActiveTab() {
    if let active = tabs.active { closeTab(active) }
  }

  /// vim's `:qa`: closes every tab, the last one first.
  func closeAllTabs() {
    for path in tabs.tabs.reversed() { closeTab(path) }
  }

  /// vim's `gt`/`gT`, `:tabnext 3`: the tab `delta` tabs away (wrapping around) or at `index`.
  func switchTab(_ to: EditorTabSwitch) {
    let list = tabs.tabs
    guard !list.isEmpty else { return }
    let current = tabs.active.flatMap { list.firstIndex(of: $0) } ?? 0
    let index: Int
    switch to {
    case .delta(let delta): index = ((current + delta) % list.count + list.count) % list.count
    case .index(let target): index = target
    }
    guard list.indices.contains(index) else { return }
    activateTab(list[index])
  }

  func closeOtherTabs(except path: String) {
    for other in tabs.tabs where other != path { closeTab(other) }
  }

  func reopenClosedTab() async {
    guard let closed = tabs.popClosedTab(isValid: { vault.isFile($0) || notes.has($0) }) else {
      return
    }
    if !notes.has(closed.path) {
      do {
        try await notes.load(closed.path)
      } catch {
        toasts.error("Couldn't reopen “\(VaultPath.stem(closed.path))”", error)
        return
      }
    }
    switchEditor(to: closed.path) { tabs.reinsert(closed) }
    afterActivate(closed.path, OpenOptions())
  }

  func goBack() async {
    guard let target = tabs.popBack(isValid: { vault.isFile($0) || notes.has($0) }) else { return }
    await openNote(target, OpenOptions(recordHistory: false))
  }

  func goForward() async {
    guard let target = tabs.popForward(isValid: { vault.isFile($0) || notes.has($0) }) else {
      return
    }
    await openNote(target, OpenOptions(recordHistory: false))
  }

  func selectTab(number: Int) {
    if let path = tabs.tab(forShortcut: number) { activateTab(path) }
  }

  func selectAdjacentTab(_ offset: Int) {
    if let path = tabs.adjacentTab(offset) { activateTab(path) }
  }

  /// Opens a `[[wikilink]]` target like Obsidian (exact path, else shortest matching stem), creating
  /// `<target>.md` when nothing matches.
  func openWikiLink(_ rawTarget: String, newTab: Bool) async {
    let target = NotePaths.wikiLinkTarget(rawTarget)
    if let resolved = WikiLinks.resolve(target, in: vault.files) {
      await openNote(resolved, OpenOptions(newTab: newTab))
      return
    }
    let normalized = VaultPath.normalize(target)
    guard !normalized.isEmpty, NotePaths.validateName(VaultPath.basename(normalized)) == nil else {
      toasts.show(.error, "Invalid link target", body: rawTarget)
      return
    }
    await createNote(
      at: VaultPath.ensureMarkdownExtension(normalized), newTab: newTab, focusTitle: false)
  }

  /// Restores last session's tabs (existing notes only), loading them in parallel. `showActive`
  /// false leaves the editor alone (another note is about to be shown).
  func restoreTabs(_ paths: [String], active: String?, showActive: Bool = true) async {
    let existing = paths.filter { vault.isFile($0) }
    guard !existing.isEmpty else { return }
    let loads = existing.filter { !notes.has($0) }.map { path in
      Task { try? await notes.load(path) }
    }
    for load in loads { await load.value }
    let loaded = existing.filter { notes.has($0) }
    guard !loaded.isEmpty else { return }
    let target = active.flatMap { loaded.contains($0) ? $0 : nil } ?? loaded[0]
    if showActive {
      switchEditor(to: target) { tabs.restore(tabs: loaded, active: target) }
      afterActivate(target, OpenOptions(focusEditor: false))
    } else {
      tabs.restore(tabs: loaded, active: target)
    }
  }

  // MARK: - Helpers

  /// Content to show for a loaded note without an editor snapshot.
  func content(of path: String) -> String? {
    notes.pendingContent(path) ?? notes.serverContent(path)
  }

  /// Saves the outgoing note, applies the tab change, and swaps the editor document.
  private func switchEditor(to path: String, _ changeTabs: () -> Void) {
    if let previous = tabs.active, previous != path { notes.saveNow(previous) }
    changeTabs()
    editor.show(path, content: content(of: path), caretAtEnd: isDailyNote(path))
  }

  /// New tasks go at the end of a daily note, so that's where its caret starts.
  func isDailyNote(_ path: String) -> Bool {
    DailyNotes.date(forPath: path, settings: settings.settings.dailyNotes) != nil
  }

  func afterActivate(_ path: String, _ options: OpenOptions) {
    if let line = options.line { editor.scrollToLine(line) }
    if options.focusEditor { editor.focus() }
    ui.expand(VaultPath.ancestorFolders(path))
    touch(path)
    refreshRecords(path)
    onTabsChanged?()
  }

  func onNoActiveNote() {
    presence.reset()
    editor.show(nil, content: nil)
  }

  /// Keeps a bounded set of recently used notes loaded (instant back-navigation).
  func touch(_ path: String) {
    recent.removeAll { $0 == path }
    recent.insert(path, at: 0)
    var index = recent.count - 1
    while recent.count > Self.maxCachedNotes, index >= Self.maxCachedNotes {
      let candidate = recent[index]
      if !tabs.tabs.contains(candidate), !notes.isBusy(candidate), candidate != tabs.active {
        recent.remove(at: index)
        notes.forget(candidate)
        editor.forget(candidate)
      }
      index -= 1
    }
  }

  func renameRecent(from: String, to: String) {
    recent = recent.map { NotePaths.renamed($0, from: from, to: to) ?? $0 }
  }

  func dropRecent(_ path: String) {
    recent.removeAll { $0 == path }
  }
}
